unstructured/ingest/connector/vectara.py

```diff
@@ -0,0 +1,246 @@
+
+    @DestinationConnectionError.wrap
+    def vectara(self):
+        """
+        Check the connection for Vectara and validate corpus exists.
+        - If more than one exists - then return a message
+        - If exactly one exists with this name - use it.
+        - If does not exist - create it.
+        """
+        try:
+            jwt_token = self._get_jwt_token()
+            if not jwt_token:
+                return "Unable to get JWT Token. Confirm your Client ID and Client Secret."
+
+            list_corpora_response = self._request(
+                endpoint="list-corpora",
+                data={"numResults": 100, "filter": self.connector_config.corpus_name},
+            )
+
+            possible_corpora_ids_names_map = {
+                corpus.get("id"): corpus.get("name")
+                for corpus in list_corpora_response.get("corpus")
+                if corpus.get("name") == self.connector_config.corpus_name
+            }
+
+            if len(possible_corpora_ids_names_map) > 1:
+                return f"Multiple Corpora exist with name {self.connector_config.corpus_name}"
+            if len(possible_corpora_ids_names_map) == 1:
+                self.connector_config.corpus_id = list(possible_corpora_ids_names_map.keys())[0]
+            else:
+                data = {
+                    "corpus": {
+                        "name": self.connector_config.corpus_name,
+                    }
+                }
+                create_corpus_response = self._request(endpoint="create-corpus", data=data)
+                self.connector_config.corpus_id = create_corpus_response.get("corpusId")
+
+        except Exception as e:
+            return str(e) + "\n" + "".join(traceback.TracebackException.from_exception(e).format())
+
+    def initialize(self):
+        self.vectara()
+
+    def _request(
+        self,
+        endpoint: str,
+        http_method: str = "POST",
+        params: t.Mapping[str, t.Any] = None,
+        data: t.Mapping[str, t.Any] = None,
+    ):
+        url = f"{self.BASE_URL}/{endpoint}"
+
+        current_ts = datetime.datetime.now().timestamp()
+        if self.jwt_token_expires_ts - current_ts <= 60:
+            self._get_jwt_token()
+
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+            "Authorization": f"Bearer {self.jwt_token}",
+            "customer-id": self.connector_config.customer_id,
+            "X-source": "unstructured",
+        }
+
+        response = requests.request(
+            method=http_method, url=url, headers=headers, params=params, data=json.dumps(data)
+        )
+        response.raise_for_status()
+        return response.json()
+
+    # get OAUth2 JWT token
+    def _get_jwt_token(self):
+        """Connect to the server and get a JWT token."""
+        token_endpoint = f"https://vectara-prod-{self.connector_config.customer_id}.auth.us-west-2.amazoncognito.com/oauth2/token"  # noqa: E501
```

Comment:
Will this endpoint always have this pattern? This looks like something we might want to expose as an option. The fact that the region is hard coded already makes me feel like it won't always be the same.

Reply:
Option created with current as default.

Reply:
Great!